jwst/master_background/expand_to_2d.py

```diff
@@ -0,0 +1,314 @@
+import math
+
+import numpy as np
+
+from .. import datamodels
+from .. assign_wcs import nirspec               # for NIRSpec IFU data
+
+def expand_to_2d(input, m_bkg_spec):
+    """Expand a 1-D background to 2-D.
+
+    Parameters
+    ----------
+    input : JWST data model
+        The input science data.
+
+    m_bkg_spec : string or JWST data model
+        Either the name of a file containing a 1-D background spectrum,
+        or a data model containing such a spectrum.
+
+    Returns
+    -------
+    background : JWST data model
+        A copy of `input` but with the data replaced by the background,
+        "expanded" from 1-D to 2-D.
+    """
+
+    with datamodels.open(m_bkg_spec) as bkg:
+        if len(bkg.spec) > 1:
+            print("Warning:  The input 1-D spectrum contains multiple spectra")
+        tab_wavelength = bkg.spec[0].spec_table['wavelength'].copy()
+        try:
+            tab_npixels = bkg.spec[0].spec_table['npixels'].copy()
+        except KeyError:
+            tab_npixels = np.ones_like(tab_wavelength)
+        tab_background = bkg.spec[0].spec_table['flux'] / tab_npixels
+
+    # We're going to use np.interp, so tab_wavelength must be strictly
+    # increasing.
+    if not np.all(np.diff(tab_wavelength) > 0):
+        index = np.argsort(tab_wavelength)
+        tab_wavelength = tab_wavelength[index].copy()
+        tab_background = tab_background[index].copy()
+
+    # Handle associations, or input ModelContainers
+    if isinstance(input, datamodels.ModelContainer):
```

Comment:
I wonder if we should let some higher-level routine, i.e. between the step class and this routine, take care of the handling of different forms of input science/source data, in which case this routine could assume it's always given a single data model instance and even a single source instance (e.g. if the input to the step is a MultiSlitModel, only pass one Slit instance at a time to this function).

Reply:
I think this *is* the intermediate function we were thinking of, as the functions this calls (i.e. `bkg_for_multislit`) actually do the work of making the background.

Reply:
This means that some of the code in the top-level `MasterBackgroundStep` class can be torn out, and we can just do the subtraction of the background in the `process()` method.

Reply:
Some of them separate cleanly, but not all.

Reply:
I can do that in a separate PR that pulls this functionality into the step.